### PR TITLE
SYNCOPE-1558: restful SAML SP metadata/keystore for WA authN delegation

### PR DESCRIPTION
**Pull request by mmoayyed**
JIRA: https://issues.apache.org/jira/browse/SYNCOPE-1558

- [X] Publish CAS SNAPSHOT to utilize latest enhancements
- [X] Design REST API endpoints to handle keystore/metadata
- [X] Provide implementation body for keystore/metadata generators in WA
- [X] Write integration tests for endpoints and services
- [X] Write unit tests for components

### CODE REVIEW

File: core/persistence-jpa/src/test/resources/domains/MasterContent.xml [hunk 1]
```diff
@@ -68,9 +68,10 @@ under the License.
   <AuthModule id="4c3ed4e6-7008-11ea-bc55-0242ac130003" name="DefaultGoogleMfaAuthModule"
```
Comment (ilgrosso) [inline core/persistence-jpa/src/test/resources/domains/MasterContent.xml:68]: You need to apply the same changes in `core/persistence-jpa-json/src/test/resources/domains/MasterContent.xml`

File: wa/starter/pom.xml [hunk 2]
```diff
@@ -237,12 +241,14 @@ under the License.
       <groupId>org.springframework.boot</groupId>
       <artifactId>spring-boot-starter-security</artifactId>
     </dependency>
-
+    <dependency>
```
Comment (ilgrosso) [inline wa/starter/pom.xml:244]: Not a big issue, but it would be nice to separate groups of dependencies by a blank line: e.g. can you add one above this?

File: wa/starter/pom.xml [hunk 3]
```diff
@@ -237,12 +241,14 @@ under the License.
...
+      <groupId>org.bouncycastle</groupId>
+      <artifactId>bcprov-jdk15on</artifactId>
+    </dependency>
     <dependency>
       <groupId>org.bouncycastle</groupId>
       <artifactId>bcpkix-jdk15on</artifactId>
     </dependency>
-
```
Comment (ilgrosso) [inline wa/starter/pom.xml:245]: Not a big issue, but it would be nice to separate groups of dependencies by a blank line: e.g. can you restore the blank line here?

File: wa/starter/src/test/java/org/apache/syncope/wa/pac4j/saml/BaseSyncopeWASAML2Client.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (ilgrosso) [inline wa/starter/src/test/java/org/apache/syncope/wa/pac4j/saml/BaseSyncopeWASAML2Client.java:1]: Can you move this package under `org.apache.syncope.wa.starter` which should be the root package of this module?

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * REST operations for SAML 2.0 SP Keystore.
+ */
+@Tag(name = "SAML 2.0 SP Keystore")
+@SecurityRequirements({
+    @SecurityRequirement(name = "BasicAuthentication"),
+    @SecurityRequirement(name = "Bearer") })
+@Path("saml2sp/conf/keystore")
+public interface SAML2SPKeystoreConfService extends JAXRSService {
```
Comment (ilgrosso) [inline common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java:47]: What it the difference to this service bound under `saml2sp/conf/keystore` and the other one bound under "saml2sp/keystore"?

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataConfService.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import javax.ws.rs.core.MediaType;
+
+/**
+ * REST operations for SAML 2.0 SP metadata.
+ */
+@Tag(name = "SAML 2.0 SP Metadata")
+@SecurityRequirements({
+    @SecurityRequirement(name = "BasicAuthentication"),
+    @SecurityRequirement(name = "Bearer")})
+@Path("saml2sp/conf/metadata")
```
Comment (ilgrosso) [inline common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataConfService.java:46]: What it the difference to this service bound under `saml2sp/conf/metadata` and the other one bound under "saml2sp/metadata"?

File: common/am/lib/src/main/java/org/apache/syncope/common/lib/types/AMEntitlement.java [hunk 7]
```diff
@@ -60,6 +60,18 @@
 
     public static final String SAML2_IDP_METADATA_READ = "SAML2_IDP_METADATA_READ";
 
+    public static final String SAML2_SP_METADATA_CREATE = "SAML2_SP_METADATA_CREATE";
```
Comment (ilgrosso) [inline common/am/lib/src/main/java/org/apache/syncope/common/lib/types/AMEntitlement.java:63]: So SAML metadata and keystore cannot be deleted?

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * Returns the SAML 2.0 SP keystore matching the given key.
+     *
+     * @param key key of requested SAML 2.0 SP keystore
+     * @return SAML 2.0 SP keystore with matching id
+     */
+    @GET
+    @Path("{key}")
+    @Produces({ MediaType.APPLICATION_JSON, RESTHeaders.APPLICATION_YAML, MediaType.APPLICATION_XML })
+    SAML2SPKeystoreTO read(@NotNull @PathParam("key") String key);
```
Comment (ilgrosso) [inline common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java:72]: What is the difference between `get()` and `read()`? I think it should be made more clear by reviewing the REST mapping.

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataService.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * Returns the SAML 2.0 SP metadata matching the given key.
+     *
+     * @param key key of requested SAML 2.0 SP metadata
+     * @return SAML 2.0 SP metadata with matching id
+     */
+    @GET
+    @Path("{key}")
+    @Produces({ MediaType.APPLICATION_JSON, RESTHeaders.APPLICATION_YAML, MediaType.APPLICATION_XML })
+    SAML2SPMetadataTO read(@NotNull @PathParam("key") String key);
```
Comment (ilgrosso) [inline common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataService.java:72]: What is the difference between `get()` and `read()`? I think it should be made more clear by reviewing the REST mapping.

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java (hunk 5 above)
Comment (mmoayyed) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java:47]: This is modeled after the SAML2 IdP metadata changes; `saml2sp/keystore` endpoints creates a keystore only if one is not found with anonymous credentials. `saml2sp/conf/keystore` allows for updates with a dedicated permission.

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataConfService.java (hunk 6 above)
Comment (mmoayyed) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataConfService.java:46]: Similar to above; `saml2sp/metadata` endpoints creates metadata only if one is not found with anonymous credentials. `saml2sp/conf/metadata` allows for updates with a dedicated permission.

File: common/am/lib/src/main/java/org/apache/syncope/common/lib/types/AMEntitlement.java (hunk 7 above)
Comment (mmoayyed) [reply common/am/lib/src/main/java/org/apache/syncope/common/lib/types/AMEntitlement.java:63]: No, it should not, or it would be catastrophic. Event updates have to be very very carefully planned.

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java (hunk 8 above)
Comment (mmoayyed) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java:72]: It might be better to rename get() with getByOwner() or some such as it attempts to locate keystore by client-name or owner really. (Similar naming convention is also used for the saml2 idp changes)

Comment (ilgrosso) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java:72]: `getByOwner()` sounds much better; in such a case, the owner value should be passed as query parameter, according to JAX-RS / REST mapping conventions, no?

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java (hunk 5 above)
Comment (ilgrosso) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java:47]: Not sure to get it: is `saml2p/keystore` specifically meant for invocation by WA? If so, its sole method should be moved in the `https://github.com/apache/syncope/tree/master/common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa` package and mapped under `wa/saml2sp/keystore`

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataConfService.java (hunk 6 above)
Comment (ilgrosso) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPMetadataConfService.java:46]: same as above

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java (hunk 8 above)
Comment (mmoayyed) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java:72]: Yes, indeed. We have that modeled it seems, 

```
    @GET
    @Produces({ MediaType.APPLICATION_JSON, RESTHeaders.APPLICATION_YAML, MediaType.APPLICATION_XML })
    SAML2SPMetadataTO get(@QueryParam("name") String name);
```

I'll proceed with the rename.

Comment (ilgrosso) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreService.java:72]: oh, sorry for overlooking that :-/

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java (hunk 5 above)
Comment (mmoayyed) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/SAML2SPKeystoreConfService.java:47]: Sure, will do.

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2IdPMetadataService.java [hunk 10]
```diff
@@ -7,7 +7,7 @@
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
- *   http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ilgrosso) [inline common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2IdPMetadataService.java:10]: Please revert this change, watch out license header re-format...

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPKeystoreService.java [hunk 11]
```diff
@@ -7,7 +7,7 @@
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
- *   http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ilgrosso) [inline common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPKeystoreService.java:10]: Same as above

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPMetadataService.java [hunk 12]
```diff
@@ -7,7 +7,7 @@
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
- *   http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ilgrosso) [inline common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPMetadataService.java:10]: Same as above

File: core/am/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/wa/SAML2IdPMetadataServiceImpl.java [hunk 13]
```diff
@@ -7,7 +7,7 @@
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
- *   http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ilgrosso) [inline core/am/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/wa/SAML2IdPMetadataServiceImpl.java:10]: Same as above

File: core/am/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/wa/SAML2SPKeystoreServiceImpl.java [hunk 14]
```diff
@@ -7,7 +7,7 @@
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
- *   http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ilgrosso) [inline core/am/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/wa/SAML2SPKeystoreServiceImpl.java:10]: Same as above

File: core/am/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/wa/SAML2SPMetadataServiceImpl.java [hunk 15]
```diff
@@ -7,7 +7,7 @@
  * "License"); you may not use this file except in compliance
  * with the License.  You may obtain a copy of the License at
  *
- *   http://www.apache.org/licenses/LICENSE-2.0
+ *    http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ilgrosso) [inline core/am/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/wa/SAML2SPMetadataServiceImpl.java:10]: v

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2IdPMetadataService.java (hunk 10 above)
Comment (ilgrosso) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2IdPMetadataService.java:10]: @mmoayyed was this overlooked by your last commit?

File: common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPKeystoreService.java (hunk 11 above)
Comment (ilgrosso) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPKeystoreService.java:10]: @mmoayyed was this overlooked by your last commit?

Comment (mmoayyed) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPKeystoreService.java:10]: I think so. I am going to pass through all components and make sure formatting is right.

Comment (mmoayyed) [reply common/am/rest-api/src/main/java/org/apache/syncope/common/rest/api/service/wa/SAML2SPKeystoreService.java]: Pushed another commit to fix formatting issues.

**Comment by ilgrosso** [comment]
@mmoayyed looks like you need to add something like as

```xml
          <testResource>
            <directory>${basedir}/../persistence-jpa/src/test/resources</directory>
            <filtering>true</filtering>
          </testResource>
```

to `persistence-jpa-json/pom.xml`

see https://travis-ci.org/github/apache/syncope/jobs/685613956#L778

File: core/persistence-jpa-json/pom.xml [hunk 16]
```diff
@@ -321,6 +331,16 @@ under the License.
           <version>${jdbc.mysql.version}</version>
           <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (ilgrosso) [inline core/persistence-jpa-json/pom.xml:334]: maybe you can add these dependencies in the main profile, to avoid duplication (there are of course no tests run by the main profile)

Comment (mmoayyed) [reply core/persistence-jpa-json/pom.xml:334]: You mean in the root pom under a <dependencies> tag, right?

Comment (ilgrosso) [reply core/persistence-jpa-json/pom.xml:334]: @mmoayyed when applying the diff in https://paste.apache.org/69xl3, I verified that both `-Ppostgres` and `-Pmysql` pass, under `core/persistence-jpa-json` (the diff also includes the fix for comment above).

Comment (mmoayyed) [reply core/persistence-jpa-json/pom.xml]: Thanks very much. Unit tests are passing. Moving onto FITs.
